### ISSUE TITLE
📈 Enhanced Monitoring - Performance & Health Analytics

### ISSUE THREAD

**Issue by timmybird** [PitchConnect/fogis-deployment#6]
## 🎯 Goal
Implement comprehensive monitoring and analytics system with real-time performance metrics, health tracking, alerting, and historical analysis for proactive system management.

## 📋 Current State
Users have:
- Basic health checks with simple up/down status
- Manual log checking for troubleshooting
- No performance metrics or trends
- No proactive alerting for issues
- Limited visibility into system behavior

## 🚀 Desired State
```bash
./manage_fogis_system.sh monitor
# Opens comprehensive monitoring dashboard
# Real-time metrics, alerts, and analytics
# Proactive issue detection and resolution
```

### Performance Monitoring
- [ ] **System Metrics** - CPU, memory, disk, network usage per service
- [ ] **Application Metrics** - Response times, throughput, error rates
- [ ] **Business Metrics** - Matches processed, calendar events synced, etc.
- [ ] **Resource Utilization** - Container resource consumption tracking
- [ ] **Performance Trends** - Historical data with trend analysis
- [ ] **Bottleneck Detection** - Identify performance bottlenecks automatically

### Health Monitoring
- [ ] **Service Health** - Advanced health checks beyond simple ping
- [ ] **Dependency Health** - Monitor external service dependencies
- [ ] **Data Quality** - Validate data integrity and consistency
- [ ] **Integration Health** - Monitor Google APIs, FOGIS connectivity
- [ ] **Synthetic Monitoring** - Automated end-to-end testing
- [ ] **Health Score** - Overall system health scoring

### Alerting System
- [ ] **Smart Alerts** - Intelligent alerting with noise reduction
- [ ] **Alert Channels** - Email, Slack, SMS, webhook notifications
- [ ] **Alert Escalation** - Escalation policies for critical issues
- [ ] **Alert Correlation** - Group related alerts to reduce noise
- [ ] **Custom Thresholds** - Configurable alert thresholds per service
- [ ] **Maintenance Windows** - Suppress alerts during maintenance

### Analytics & Reporting
- [ ] **Performance Reports** - Automated daily/weekly/monthly reports
- [ ] **Capacity Planning** - Resource usage forecasting
- [ ] **SLA Monitoring** - Track service level agreements
- [ ] **Error Analysis** - Detailed error tracking and categorization
- [ ] **Usage Analytics** - User behavior and system usage patterns
- [ ] **Cost Analysis** - Resource cost tracking and optimization

### Phase 1: Core Monitoring Infrastructure
```yaml
# docker-compose.monitoring.yml
services:
  prometheus:
    image: prom/prometheus:latest
    ports:
      - "9090:9090"
    volumes:
      - ./monitoring/prometheus.yml:/etc/prometheus/prometheus.yml
      - prometheus_data:/prometheus
    command:
      - '--config.file=/etc/prometheus/prometheus.yml'
      - '--storage.tsdb.path=/prometheus'
      - '--web.console.libraries=/etc/prometheus/console_libraries'
      - '--web.console.templates=/etc/prometheus/consoles'
      - '--storage.tsdb.retention.time=30d'
      - '--web.enable-lifecycle'

  grafana:
    image: grafana/grafana:latest
    ports:
      - "3000:3000"
    environment:
      - GF_SECURITY_ADMIN_PASSWORD=admin
    volumes:
      - grafana_data:/var/lib/grafana
      - ./monitoring/grafana/dashboards:/etc/grafana/provisioning/dashboards
      - ./monitoring/grafana/datasources:/etc/grafana/provisioning/datasources

  alertmanager:
    image: prom/alertmanager:latest
    ports:
      - "9093:9093"
    volumes:
      - ./monitoring/alertmanager.yml:/etc/alertmanager/alertmanager.yml
      - alertmanager_data:/alertmanager

  node-exporter:
    image: prom/node-exporter:latest
    ports:
      - "9100:9100"
    volumes:
      - /proc:/host/proc:ro
      - /sys:/host/sys:ro
      - /:/rootfs:ro
    command:
      - '--path.procfs=/host/proc'
      - '--path.rootfs=/rootfs'
      - '--path.sysfs=/host/sys'
      - '--collector.filesystem.mount-points-exclude=^/(sys|proc|dev|host|etc)($$|/)'
```

### Phase 2: Application Metrics
```python
# lib/metrics_collector.py
from prometheus_client import Counter, Histogram, Gauge, start_http_server

class FogisMetrics:
    def __init__(self):
        # Business metrics
        self.matches_processed = Counter(
            'fogis_matches_processed_total',
            'Total number of matches processed'
        )
        
        self.calendar_events_synced = Counter(
            'fogis_calendar_events_synced_total',
            'Total calendar events synced'
        )
        
        # Performance metrics
        self.api_request_duration = Histogram(
            'fogis_api_request_duration_seconds',
            'Time spent on API requests',
            ['endpoint', 'method']
        )
        
        self.active_connections = Gauge(
            'fogis_active_connections',
            'Number of active connections'
        )
        
        # Health metrics
        self.service_health = Gauge(
            'fogis_service_health',
            'Service health status (1=healthy, 0=unhealthy)',
            ['service']
        )
    
    def record_match_processed(self):
        self.matches_processed.inc()
    
    def record_api_request(self, endpoint, method, duration):
        self.api_request_duration.labels(
            endpoint=endpoint, method=method
        ).observe(duration)
```

### Phase 3: Advanced Analytics
- Machine learning for anomaly detection
- Predictive analytics for capacity planning
- Advanced correlation analysis

### System Overview Dashboard
```json
{
  "dashboard": {
    "title": "FOGIS System Overview",
    "panels": [
      {
        "title": "Service Health Status",
        "type": "stat",
        "targets": [
          {
            "expr": "fogis_service_health",
            "legendFormat": "{{service}}"
          }
        ]
      },
      {
        "title": "Matches Processed (24h)",
        "type": "stat",
        "targets": [
          {
            "expr": "increase(fogis_matches_processed_total[24h])"
          }
        ]
      },
      {
        "title": "API Response Times",
        "type": "graph",
        "targets": [
          {
            "expr": "histogram_quantile(0.95, fogis_api_request_duration_seconds_bucket)",
            "legendFormat": "95th percentile"
          }
        ]
      },
      {
        "title": "Resource Usage",
        "type": "graph",
        "targets": [
          {
            "expr": "rate(container_cpu_usage_seconds_total[5m]) * 100",
            "legendFormat": "CPU %"
          },
          {
            "expr": "container_memory_usage_bytes / container_spec_memory_limit_bytes * 100",
            "legendFormat": "Memory %"
          }
        ]
      }
    ]
  }
}
```

### Business Metrics Dashboard
```json
{
  "dashboard": {
    "title": "FOGIS Business Metrics",
    "panels": [
      {
        "title": "Match Processing Rate",
        "type": "graph",
        "targets": [
          {
            "expr": "rate(fogis_matches_processed_total[1h])",
            "legendFormat": "Matches/hour"
          }
        ]
      },
      {
        "title": "Calendar Sync Success Rate",
        "type": "stat",
        "targets": [
          {
            "expr": "(fogis_calendar_events_synced_total / fogis_calendar_events_attempted_total) * 100"
          }
        ]
      },
      {
        "title": "Error Rate by Service",
        "type": "graph",
        "targets": [
          {
            "expr": "rate(fogis_errors_total[5m])",
            "legendFormat": "{{service}}"
          }
        ]
      }
    ]
  }
}
```

### Critical Alerts
```yaml
# monitoring/alerts/critical.yml
groups:
  - name: fogis.critical
    rules:
      - alert: ServiceDown
        expr: fogis_service_health == 0
        for: 1m
        labels:
          severity: critical
        annotations:
          summary: "FOGIS service {{ $labels.service }} is down"
          description: "Service {{ $labels.service }} has been down for more than 1 minute"
      
      - alert: HighErrorRate
        expr: rate(fogis_errors_total[5m]) > 0.1
        for: 2m
        labels:
          severity: critical
        annotations:
          summary: "High error rate detected"
          description: "Error rate is {{ $value }} errors/second for service {{ $labels.service }}"
      
      - alert: APIResponseTimeTooHigh
        expr: histogram_quantile(0.95, fogis_api_request_duration_seconds_bucket) > 5
        for: 5m
        labels:
          severity: warning
        annotations:
          summary: "API response time too high"
          description: "95th percentile response time is {{ $value }}s"
```

### Resource Alerts
```yaml
# monitoring/alerts/resources.yml
groups:
  - name: fogis.resources
    rules:
      - alert: HighCPUUsage
        expr: rate(container_cpu_usage_seconds_total[5m]) * 100 > 80
        for: 5m
        labels:
          severity: warning
        annotations:
          summary: "High CPU usage detected"
          description: "CPU usage is {{ $value }}% for container {{ $labels.name }}"
      
      - alert: HighMemoryUsage
        expr: (container_memory_usage_bytes / container_spec_memory_limit_bytes) * 100 > 85
        for: 5m
        labels:
          severity: warning
        annotations:
          summary: "High memory usage detected"
          description: "Memory usage is {{ $value }}% for container {{ $labels.name }}"
      
      - alert: DiskSpaceLow
        expr: (node_filesystem_avail_bytes / node_filesystem_size_bytes) * 100 < 10
        for: 1m
        labels:
          severity: critical
        annotations:
          summary: "Disk space critically low"
          description: "Only {{ $value }}% disk space remaining"
```

### Basic Monitoring Operations
```bash
# Start monitoring stack
./manage_fogis_system.sh monitor start
# 📈 Starting monitoring services...
# ✅ Prometheus: http://localhost:9090
# ✅ Grafana: http://localhost:3000 (admin/admin)
# ✅ AlertManager: http://localhost:9093

# Check monitoring status
./manage_fogis_system.sh monitor status
# 📈 Monitoring Status
# ==================
# ✅ Prometheus: Healthy (collecting metrics from 6 services)
# ✅ Grafana: Healthy (3 dashboards configured)
# ✅ AlertManager: Healthy (2 active alerts)
# 📊 Data retention: 30 days

# View current alerts
./manage_fogis_system.sh monitor alerts
# 🚨 Active Alerts
# ===============
# ⚠️ WARNING: High CPU usage on match-list-processor (85%)
# 🔴 CRITICAL: Service down - team-logo-combiner

# Generate performance report
./manage_fogis_system.sh monitor report --period=7d
# 📈 FOGIS Performance Report (Last 7 days)
# ============================================
# Uptime: 99.2%
# Matches processed: 1,247
# Average response time: 245ms
# Error rate: 0.03%
```

### Advanced Analytics
```bash
# Capacity planning analysis
./manage_fogis_system.sh monitor capacity
# 📈 Capacity Analysis
# ==================
# Current resource usage: 65%
# Projected usage (30 days): 78%
# Recommended action: Scale up in 3 weeks

# Performance optimization suggestions
./manage_fogis_system.sh monitor optimize
# 🚀 Optimization Suggestions
# ============================
# 1. match-list-processor: Consider caching (30% performance gain)
# 2. fogis-api-client: Reduce connection pool size (15% memory savings)
# 3. team-logo-combiner: Optimize image processing (40% faster)
```

## 🧪 Testing Requirements
- [ ] **Metrics Collection** - Verify all metrics are collected correctly
- [ ] **Alert Testing** - Test alert firing and resolution
- [ ] **Dashboard Functionality** - Verify all dashboard panels work
- [ ] **Performance Impact** - Ensure monitoring doesn't impact system performance
- [ ] **Data Retention** - Test data retention and cleanup
- [ ] **Failover Testing** - Test monitoring system resilience

## 📊 Success Metrics
- 99.9% monitoring system uptime
- <1% performance overhead from monitoring
- 95% alert accuracy (low false positives)
- <5 minute mean time to detection (MTTD)
- 90% reduction in manual system checking

## 🔗 Dependencies
- Prometheus for metrics collection
- Grafana for visualization
- AlertManager for alerting
- Service instrumentation for metrics

### Machine Learning Anomaly Detection
```python
# lib/anomaly_detection.py
class AnomalyDetector:
    def __init__(self):
        self.model = IsolationForest(contamination=0.1)
    
    def detect_anomalies(self, metrics_data):
        """Detect anomalies in system metrics"""
        anomalies = self.model.fit_predict(metrics_data)
        return self.format_anomaly_alerts(anomalies)
```

### Predictive Analytics
```python
# lib/predictive_analytics.py
class CapacityPredictor:
    def predict_resource_needs(self, historical_data, forecast_days=30):
        """Predict future resource requirements"""
        model = LinearRegression()
        model.fit(historical_data)
        return model.predict(forecast_days)
```

### Custom Metrics
```bash
# Add custom business metrics
./manage_fogis_system.sh monitor add-metric \
  --name="referee_assignments" \
  --type="counter" \
  --description="Number of referee assignments made"
```

---
**Priority**: 🔥 MEDIUM - Improves operational visibility
**Effort**: 📊 HIGH - 1-2 weeks development
**Impact**: 🚀 HIGH - Enables proactive system management
